### PR TITLE
Record push information in PushRun structure

### PR DESCRIPTION
**Pull request by garionpin**
Record the TextUnits that were associated with a specific PushRun in the
database to aid state-based delta generation.

### CODE REVIEW

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PushCommand.java [hunk 1]
```diff
@@ -89,7 +100,43 @@ public void execute() throws CommandException {
...
+
+        if (recordPushRun && commitHash == null) {
+            throw new IllegalArgumentException(
+                    MessageFormat.format(
+                            "Whenever {0} is set to true, {1} must also be passed in with a commit hash.",
+                            Param.RECORD_PUSH_RUN_LONG,
+                            Param.COMMIT_HASH_LONG));
+        }
+
+        if (recordPushRun && !StringUtils.isWhitespace(commitHash) && pushType == PushService.PushType.SEND_ASSET_NO_WAIT_NO_DELETE) {
```
Comment (baktayPin) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/PushCommand.java:126]: Do we need the !StringUtils.isWhitespace(commitHash) check here as it's contained in the first check already?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/param/Param.java [hunk 2]
```diff
@@ -147,4 +147,13 @@ public class Param {
...
     public static final String PUSH_TYPE_DESCRIPTION = "To choose the push type. Don't change unless you know exactly what it does";
+
+    public static final String COMMIT_HASH_LONG = "--commit-hash";
+    public static final String COMMIT_HASH_SHORT = "-c";
+    public static final String COMMIT_HASH_DESCRIPTION = "The commit hash that will be used when recording a push-run. Used together with --record-push-run.";
+
+
+    public static final String RECORD_PUSH_RUN_LONG = "--record-push-run";
+    public static final String RECORD_PUSH_RUN_SHORT = "-rp";
+    public static final String RECORD_PUSH_RUN_DESCRIPTION = "Stores a full list of text units that were processed by the specified commit, through an abstraction called `push run`";
```
Comment (baktayPin) [inline cli/src/main/java/com/box/l10n/mojito/cli/command/param/Param.java:158]: Isn't this parameter description supposed to describe a bool value? Should the description be updated here?

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PushCommand.java (hunk 1 above)
Comment (garionpin) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PushCommand.java:126]: Indeed, removed it for the next iteration.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/param/Param.java (hunk 2 above)
Comment (garionpin) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/param/Param.java:158]: This parameter doesn't actually take in a boolean parameter - it's only passed as is, I've clarified the description in the next iteration.

File: cli/src/main/java/com/box/l10n/mojito/cli/command/PushCommand.java [hunk 3]
```diff
@@ -89,7 +100,43 @@ public void execute() throws CommandException {
 
         Repository repository = commandHelper.findRepositoryByName(repositoryParam);
 
-        ArrayList<FileMatch> sourceFileMatches = commandHelper.getSourceFileMatches(commandDirectories, fileType, sourceLocale, sourcePathFilterRegex);
+        if (commitHash != null && StringUtils.isBlank(commitHash)) {
+            throw new IllegalArgumentException(
```
[the remark this answers is not in the document]
Comment (garionpin) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/PushCommand.java:104]: Done in the next iteration.

File: webapp/src/main/java/com/box/l10n/mojito/rest/pushrun/PushRunWS.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    RepositoryRepository repositoryRepository;
+
+    public PushRunWS(PushRunService pushRunService, RepositoryRepository repositoryRepository) {
+        this.pushRunService = pushRunService;
+        this.repositoryRepository = repositoryRepository;
+    }
+
+    @JsonView(View.PushRun.class)
+    @RequestMapping(value = "/api/pushruns", method = RequestMethod.POST)
+    PushRun createPushRun(@RequestBody long repositoryId) {
```
[the remark this answers is not in the document]
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/rest/pushrun/PushRunWS.java]: Removed the WS to further simplify the flow.

File: webapp/src/main/java/com/box/l10n/mojito/rest/pushrun/PushRunWS.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        Repository repository = repositoryRepository.findById(repositoryId)
+                .orElseThrow(() -> new RuntimeException(
+                        String.format("No repository could be found for id: %s", repositoryId)));
```
[the remark this answers is not in the document]
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/rest/pushrun/PushRunWS.java]: Removed the WS to further simplify the flow.

File: webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java [hunk 6]
```diff
@@ -196,77 +200,136 @@ public class AssetExtractionService {
...
+                                                                                    filterConfigIdOverride,
+                                                                                    filterOptions,
+                                                                                    currentTask);
+        CreateTextUnitsResult createdTextUnitsResult = createTextUnitsForNewContent(assetContent,
+                                                                                    stateForNewContent,
+                                                                                    currentTask);
 
         updateBranchAssetExtraction(assetContent, createdTextUnitsResult.getUpdatedState(), filterOptions, currentTask);
         updateLastSuccessfulAssetExtraction(asset, createdTextUnitsResult.getUpdatedState(), currentTask);
+        updatePushRun(asset, createdTextUnitsResult.getUpdatedState(), pushRunId, currentTask);
```
Comment (garionpin) [inline webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java:229]: @aurambaj  Is
`createdTextUnitsResult.getUpdatedState() `
the right TU source to use here - that won't pull in TUs from other branches, but will still cover existing + new TUs?

Comment (aurambaj) [reply webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java:229]: yes, exact

File: webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java [hunk 7]
```diff
@@ -196,77 +200,136 @@ public class AssetExtractionService {
...
+                                       AssetContentMd5s.of());
+    }
+
+    @Pollable(message = "Updating push run")
+    void updatePushRun(Asset asset,
+                       MultiBranchState multiBranchState,
+                       Long pushRunId,
+                       @ParentTask PollableTask parentTask) {
+
+        if (pushRunId == null || pushRunId == 0) {
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java:253]: why `pushRunId == 0`?

File: webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java [hunk 8]
```diff
@@ -466,7 +558,8 @@ Modifications getModifications(MultiBranchState base, MultiBranchState current)
 
         // we look into entries that are used in both, and we try to update it
 
-        ImmutableMap<String, Map.Entry<String, BranchData>> baseBothUsedBranchDataByMd5 = base.getBranchStateTextUnits().stream()
+        ImmutableMap<String, Map.Entry<String, BranchData>> baseBothUsedBranchDataByMd5 = base.getBranchStateTextUnits()
```
Comment (aurambaj) [inline webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java:561]: A lot of change seems to be indentation, can we keep it out this commit for ease of reviewing?

File: webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java (hunk 6 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java:229]: Thanks for confirming!

File: webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java (hunk 7 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java:253]: AutoIncrement in SQL starts at 1 and thus 0 is always guaranteed to be an incorrect value - likely passed in for a badly initialized param (same as null effectively).

File: cli/src/main/java/com/box/l10n/mojito/cli/command/CommitCreateCommand.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+import org.springframework.context.annotation.Scope;
+import org.springframework.stereotype.Component;
+
+/**
+ * @author garion
+ */
+@Component
+@Scope("prototype")
+@Parameters(commandNames = {"commit-create", "cc"}, commandDescription = "Create commit information in Mojito.")
+public class CommitCreateCommand extends Command {
```
[the remark this answers is not in the document]
Comment (garionpin) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/CommitCreateCommand.java]: Removed from this commit. Won't attempt a stacked PR as the experience in GH atm isn't great for that. Will raise once this is merged.

File: webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java (hunk 8 above)
Comment (garionpin) [reply webapp/src/main/java/com/box/l10n/mojito/service/assetExtraction/AssetExtractionService.java:497]: Done. IntelliJ's re-format file sabotaged me there :) .

File: cli/src/main/java/com/box/l10n/mojito/cli/command/CommitCreateCommand.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (aurambaj) [reply cli/src/main/java/com/box/l10n/mojito/cli/command/CommitCreateCommand.java]: you can have multiple commits in one PR and apply fixups as you update it. It is a bit cumbersome but work okay

**Pull request merged**